### PR TITLE
Bazel files for building asterius binaries

### PR DESCRIPTION
**Pull request by ylecornec**
In order to use asterius with rules_haskell, I set up (some time ago) a bazel build of asterius that produces a self contained archive (Available [here](https://github.com/tweag/asterius_bundles_for_bazel/releases)).
This archive contains asterius and needed dependencies (the wasi-sdk fork, dynamic libraries, etc.), so that it can be easily used by rules_haskell both on nixOS and linux bindists.

Now that the new build of asterius was merged into master, the build of this bundle is not up to date. However, I started looking into it and it seems not too far from being usable.
Because of all the progress on the nix part, building the bundle is now way simpler and could run in the CI (provided that the big dependencies are in the cache already, as with the regular nix-build).

It is unclear to me how the nix/cabal build of asterius will evolve, but it may be worth merging the Bazel files that work at the moment to help keep things in sync.

## The Bazel files

This PR adds files for building the asterius binaries with bazel in the CI (but not yet the boot process and the generation of the bundle).

- The `WORKSPACE` file sets up external dependencies.
- The `BUILD.bazel` files make their directories bazel packages and contain the definitions of targets to build.
- The rest of the files will be in the `bazel` directory.

## Bazel's external dependencies

The `bazel/bazel.md` file contains the list of external dependencies that should be kept in sync between the bazel build and the cabal one:

- The `bazel/stack.yaml` file should be kept in sync with the `nix.cabal.project` files as we rely on this custom stack snapshot for external haskell dependencies.
On updating this file, the `bazel/stackage_snapshot.json` should be generated again by executing `bazel run @stackage-unpinned//:pin`.

- The `HASKELL_BINARYEN_COMMIT` variable from the `WORKSPACE` file.
(Because of a `rules_haskell` [issue](https://github.com/tweag/rules_haskell/issues/1676) we do not recover this library through the stack snapshot yet.)

- The [bazel/nix/bazel_deps.nix](https://github.com/tweag/asterius/compare/ylecornec/bazel_workspace?expand=1#diff-b900cd74fb0cce9cd8cb9bf42b16a1f47b0dc6898f794bcf5eec5309159f5fed) file mimics the `shell.nix` file and exposes necessary attributes to bazel.
If the other nix files are refactored it may be necessary to also modify this one.

### CODE REVIEW

**Review by TerrorJack — CHANGES_REQUESTED** [review]
I know little about bazel, so other than some minor points I don't see any immediate questions here.

File: bazel/nix/bazel_build_shell.nix [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{ pkgs ? import ./bazel_deps.nix {} }:
+with pkgs;
+mkShell {
+  LANG="C.UTF-8";
+  buildInputs = [gcc cacert git nix bazel_4 ];
```
[the remark this answers is not in the document]
Comment (ylecornec) [reply bazel/nix/bazel_build_shell.nix:5]: Yes, bazel does use nix to setup external dependencies. This is done in the `WORKSPACE` file by the rules starting by `nixpkgs`.

Indeed the plan was to use a pure nix shell (as in the rules_haskell setup) and I forgot to set the option. I tried adding it in a new commit and things seem to work fine.

In particular we can see [here](https://github.com/tweag/asterius/runs/4854346321?check_suite_focus=true#step:7:346) that cachix is used when bazel calls nix.

File: .github/workflows/shell.yml [hunk 2]
```diff
@@ -40,5 +54,14 @@ jobs:
...
 
       - name: shell
+        if: ${{ matrix.build-tool == 'cabal' }}
         run: |
           nix-shell --keep-going --run "cabal build asterius && asterius/boot.sh"
+
+      - name: Bazel build
+        if: ${{ matrix.build-tool == 'bazel' }}
+        run: |
+          nix-shell bazel/nix/bazel_build_shell.nix --run '
```
[the remark this answers is not in the document]
Comment (ylecornec) [reply .github/workflows/shell.yml]: I created a new file for the bazel pipeline :+1:

File: WORKSPACE [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    deps = [
+        "libbinaryen",
+        "base",
+    ],
+    visibility = ["//visibility:public"],
+    )
+""",
+    commit = HASKELL_BINARYEN_COMMIT,
+    remote = "https://github.com/tweag/haskell-binaryen",
+    shallow_since = "1610373671 +0100",
```
[the remark this answers is not in the document]
Comment (ylecornec) [reply WORKSPACE:205]: It is not required for things to work. If the history is shallow then its main use is to silence a bazel warning.

**Review by TerrorJack — APPROVED** [review]

**Pull request merged**